enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+
+In the enhancement proposal we will describe an architecture which would allow
+customers to install individual compomnents as rpm packages.
+
+
+## Motivation
+
+The principal factor that gave birth to the idea of MicroShift was the need to bring
+the OpenShift deployment model into the Edge where the power, memory and CPU budgets
+are reduced.
+
+While the default configuration makes sense to most of our customers in the Device
+Edge area, still our current CSI, CNI, and router components use 35% of the memory
+footprint and a significative CPU and power footprint. Making those components
+pluggable would allow customers who don't require advanced network features,
+persistent volumes, or ingress/routes to disable or switch those technologies
+and reduce the memory/cpu and disk/over-the-wire footprint.
+
+
+### User Stories
+
+An application developer needs the whole stack of components, creates an image
+that contains:
+1. microshift
+
+Nothing would change in this case, installing microshift, which becomnes a virtual
+package, installs the necessary dependencies:
+* microshift-core
+* microshift-networking-ovn
+* microshift-storage-topolvm
+* microshift-routes
+
+as well as any other dependencies.
+
+An application developer only needs advanced networking, but no persistent volumes,
+and no routes/ingress.  Creates an image that contains:
+1. microshift-core
+2. microshift-networking-ovn
+
+
+An application developer wants the bare minimum set of components creates an image
+with:
+1. microshift-core
+
+An application developer wanting bare minimum networking but persistent volumes
+creates an image with:
+1. microshift-core
+2. microshift-storage-topolvm
+
+### Goals
+* Allow networking, storage and router core components to be optionally
+  installed.
+
+* As we refactor and re-use the kustomize package used for loading user workloads,
+  add support for loading yamls from multiple directories, i.e.:
+  * `/usr/lib/microshift/manifests.d/user-app-1/`
+  * `/usr/lib/microshift/manifests.d/user-app-2/`
+  * `/etc/microshift/manifests.d/user-app-3/`
+  * `...`
+
+* Split the RPM packages to resemble those optional components.
+
+* Preserve previous behavior in terms of RPM package installation.
+
+### Non-Goals
+
+* Making DNS component optional.
+* Making the certificate management component optional.
+
+## Proposal
+
+To allow modularity of the described core components, the following changes
+would be introduced:
+
+Update the service manager to allow core components as kustomization yamls,
+in the same way we already support for user application manifests. i.e.:
+
+`/usr/lib/microshift/components.d/networking-ovn/kustomize.yaml`
+`/usr/lib/microshift/components.d/storage-topolvm/kustomize.yaml`
+`/usr/lib/microshift/components.d/openshift-router/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory and extract metadata from the yamls to analyze dependencies between
```

Comment:
Today we already have this problem if one core component would not boot, I guess if it runs into an error MicroShift simply crashes with an error, but we could maybe enhance the reporting of the service manager in a way that users can understand what is being booted and what is waiting on what.

Reply:
We'd want a "fail fast" approach; verify things are in place and die if not early, before services started and produced lots of logs.

Reply:
My understanding is that kustomize returns error in case manifests are in wrong format, the question is that is kustomize approach capable of validating the readiness of plugin w/o adding logic in microshift main process?

Reply:
But we don't even do that today for CNI/CSI. It's crio and the pod readiness settings what will consider the readiness of the network component itself once installed.

Reply:
We validate the prerequisite for ovnk when initOVS is disabled in config.yaml.